src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -97,13 +103,53 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     }
     String query = encodeValue(param);
     List<Card> allCards = new ArrayList<>();
-    for (Client client : clients) {
-      String link = getLink(client.getCseId(), query);
-      Card card = client.search(link);
-      if (card != null) {
-        allCards.add(card);
-      }
+    ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
+    List<Callable<Card>> cardCallbacks =
```

Comment:
So there's a lot of duplicated code here. My advice (warning: I only somewhat tried compiling this):

```java
List<Callable<Card>> cardCallbacks = clients.stream().map(client -> ((Callable<Card>)() -> {
  String link = getLink(client.getCseId(), query);
  if (link != null) {
    return client.search(link);
  }
  return null;
})).collect(Collectors.toList());
```

This maps the list of clients to a list of lambdas that return cards.

Then you can avoid mutating allCards, you can go all-in with functional programming:

```java
List<Card> cards;
try {
  cards = executor.invokeAll(...).stream().map(...).filter(card -> card != null).collect(Collectors.toList());
} catch (InterruptedException | RejectedExecutionException e) {
  cards = Collections.emptyList();
}
```

Reply:
Gotcha, still a little unfamiliar with functional programming, I'll try it out.